src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+             neighbors::AbstractVector{NTuple}) -> maxtree
+
+Rebuilds the `maxtree` for the `image` using `neighbors` as the pixel
+connectivity specification.
+
+# Arguments
+- `neighbors`: the vector of the offsets (``o_i``) to the cartesian index of
+   the pixel (``p``), so that ``p + o_i`` is the cartesian index of the ``i``-th
+   neighbor of ``p``.
+
+# Details
+The pixels in the connected components generated by the method should be
+connected to each other by a path through neighboring pixels (as defined by
+`neighbors`).
+
+# See also
+[`MaxTree`](@ref)
+"""
+function rebuild!(maxtree::MaxTree{N},
+                  image::AbstractArray{<:Number, N},
+                  #=mask::AbstractArray{Bool, N},=#
+                  neighbors::AbstractVector{NTuple{N, Int}}) where N
+    # pixels need to be sorted according to their gray level.
+    sortperm!(maxtree.traverse, vec(image), rev=maxtree.rev)
+    # some sufficiently inner point
+    center = CartesianIndex{N}(ntuple(i -> 1 + size(image, i) ÷ 2, Val{N}()))
+    # offsets for linear indices (note that they are sorted in ascending order)
+    neighbor_offsets = getindex.(Ref(LinearIndices(image)),
+        [CartesianIndex{N}(ntuple(i -> center[i] + n[i], Val{N}())) for n in neighbors]) .-
+        getindex(LinearIndices(image), center)
+
+    # initialization of the image parent
+    parents = fill!(maxtree.parents, 0) |> vec
+    roots = fill!(similar(parents), 0) # temporary array containing current roots
+    rootpath = Vector{Int}() # temp array to hold the path to the current root within ancestors
+    cindexes = CartesianIndices(maxtree.parents)
+
+    # traverse the array in reversed order (from highest value to lowest value)
+    @inbounds for p in Iterators.Reverse(maxtree.traverse)
+        p_ci = cindexes[p]
+        parents[p] = roots[p] = p # it's a new root
+
+        for i in eachindex(neighbors)
+            # check that p is in the mask and that it's neighbor is a valid image pixel
+            (#=mask[p] && =#isvalid_offset(neighbors[i], p_ci, maxtree.axes)) || continue
+            index = p + neighbor_offsets[i] # linear index of the neighbor
+            (parents[index] == 0) && continue # ignore neighbor without parent (= its value is lower)
+            root = rootpath!(rootpath, roots, index) # neighbor's root
+            if (root != p) || (length(rootpath) > 1) # attach neighbor's root to the p
+                parents[root] = p
+                roots[rootpath] .= p # also attach the ancestral nodes of neighbor to p
+            end
+        end
+    end
+    return canonize!(maxtree, image)
+end
+
+# generates a vector of offsets to the CartesianIndex that defines
+# the neighborhood of the pixel in an N-dimensional image.
+# `connectivity` is the maximal number of orthogonal steps
+# (+1/-1 to a single dimension) that are required to reach a neighbor.
+function neighbor_cartesian_offsets(::Type{CartesianIndex{N}}, connectivity::Integer) where N
```

Comment:
Dispatching on types isn't really useful here, instead,`neighbor_cartesian_offsets(dims, connectivity)` should be sufficient. Corrects me if I understand it wrongly.

Reply:
I think that without dispatching on types Julia would not be able to infer the type of the returned value.
Maybe it's not so important now, since `rebuild!()` (which gets the result of this method) would know the type from the beginning, so the type inference in the performance-critical `rebuild!()` should not be affected.
But I'd rather keep the method signature as is to ensure correct type inference if/when this method would be called in different contexts.

Reply:
Hmmm, I'll see this as an over-optimization.

If you insist, I suggest you add some `@inferred` test for this function to ensure others don't break it in the future. Also, I think dispatching on `Val` is more readable :)